Add functionality and feature flag to allow hard deleting organizations

miveronese opened the pull request:
Ticket: https://sourcegraph.atlassian.net/browse/CLOUD-81

This PR adds the following:
- a new graphql mutation to allow hard deletion of organizations 
- a new HardDelete method (in the OrgStore)
- unit tests

## Test plan
### Unit tests
Run the new tests with:
`go test ./cmd/frontend/graphqlbackend/. -v -run TestRemoveOrganization` 
and 
`go test ./internal/database/. -run  TestDeleteOrganization_OnPremise -v`
`go test ./internal/database/. -run  TestDeleteOrganization_OnCloud -v`

### Manual test (via api/console)

1. Case one: you are in the DotCom mode but don't have the feature flag "org-deletion" enabled or it is set to false
Run `sg start dotcom` 
Go to the /api/console
Try to delete an organization by running the "removeOrganization" mutation
You should see an error message informing that this operation is not supported

2. Case two: you are in the DotCom mode and the feature flag "org-deletion" is set to true
Run `sg start dotcom` 
Go to the /api/console
Try to hard delete an organization by running the "removeOrganization" mutation
All should work and the org will be hard deleted from the orgs table and from the associated tables

3. Case three: you are not in the DotCom mode
Run `sg start` 
Go to the /api/console
Try to hard delete an organization
You should see an error message informing that this operation is not supported


### Queries
````
mutation CreateFeatureFlag{
	createFeatureFlag(
      name: "org-deletion",
	  value: false,
   ){
     __typename
   }
 }
````

```
mutation {
  removeOrganization(organization: "org-id") {
        alwaysNil

  }
}
```
<!--
  As part of SOC2/GN-104 and SOC2/GN-105 requirements, all pull requests are REQUIRED to
  provide a "test plan". A test plan is a loose explanation of what you have done or
  implemented to test this, as outlined in our Testing principles and guidelines:
  https://docs.sourcegraph.com/dev/background-information/testing_principles
  Write your test plan here after the "## Test plan" header.
-->

miveronese commented on client/web/src/org/backend.ts:
```diff
@@ -127,3 +159,4 @@ export const GET_ORG_FEATURE_FLAG_VALUE = gql`
 export const ORG_CODE_FEATURE_FLAG_NAME = 'org-code'
 export const ORG_CODE_FEATURE_FLAG_EMAIL_INVITE = 'org-email-invites'
 export const GITHUB_APP_FEATURE_FLAG_NAME = 'github-app-cloud'
+export const ORG_DELETION_FEATURE_FLAG_NAME = 'org-deletion'
```
I'm  uncertain on how to properly feature flags for this case, so the  idea was to check if an org has the  "org-deletion" flag enabled and follow the example from email invitations.  

I didn't use "experimentalFeatures" because that  looks more like an option to manage external user access to some features via their user settings or organization settings that can be managed in the UI (correct me, please, if I got this wrong)

kopancek replied:
I think using the internal feature flagging API like you did is the best option here. But I think using a global flag instead of an organization override will be better.

miveronese replied:
@kopancek , do you have an example to share on the use of global flag vs org override flag? I am not sure if I got this correctly.

kopancek replied:
Yes, you can try doing just that. If that does not work, you can use the same function as you do now, but from the DB package instead: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/feature_flags.go?L560

miveronese commented on internal/database/orgs.go:
```diff
@@ -308,3 +309,44 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+	}
+
+	return nil
+}
```
@novoselrok , for notebooks, so far, I could see only the "settings" table being affected...  I'm wondering if , when I creating a notebook, I missed some step that could mean that data would be added to another table besides "settings"...

novoselrok replied:
On Cloud, users don't have to add `showSearchNotebook` to their settings. The `showSearchNotebook` feature is enabled globally in the site-admin settings.

When creating a notebook only the `notebooks` table is affected. Here is the DB [code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/notebooks/store.go?L334-359) for creating a notebook.

kopancek commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +285,27 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
 }
 
+// HardDeleteOrganization removes the org and all resources associated with this org when flag is enabled.
+func (r *schemaResolver) HardDeleteOrganization(ctx context.Context, args *struct {
```
I would call it just `removeOrganization` to follow the convention we use on other graphql resolvers. And the resolver could either do a soft delete or hard delete, depending on the feature flag value. WDYT?

miveronese replied:
That's an interesting idea, Milan! And like it and will try that approach.

kopancek replied:
But now that I think about it. If we want this feature disabled everywhere else other than cloud, it might be better if it returns an error if the feature flag value is `false`.

miveronese replied:
@kopancek , I think I need your 👍  before be allowed to merge this PR. :)

kopancek commented on internal/gqltestutil/organization.go:
```diff
@@ -202,6 +207,29 @@ mutation DeleteOrganization($organization: ID!) {
 	return nil
 }
 
+// // HardDeleteOrganization hard deletes the organization by given GraphQL node ID.
+// //
+// // This method requires the authenticated user to be a site admin.
+// func (c *Client) HardDeleteOrganization(id string) error {
+// 	const query = `
+// mutation HardDeleteOrganization($organization: ID!) {
+// 	hardDeleteOrganization(organization: $organization) {
+// 		alwaysNil
+// 	}
+// }
+// `
+// 	fmt.Println("------------ testing hard delete ORG ----------------")
+
+// 	variables := map[string]interface{}{
+// 		"organization": id,
+// 	}
+// 	err := c.GraphQL("", query, variables, nil)
+// 	if err != nil {
+// 		return errors.Wrap(err, "request GraphQL")
+// 	}
+// 	return nil
+// }
+
```
Was this intentional?

miveronese replied:
Oops, that was unintentional -- a leftover that shouldn't have been committed.

kopancek commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +285,27 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
+		return nil, err
+	}
+
+	enabled, _ := r.db.FeatureFlags().GetOrgFeatureFlag(ctx, orgID, OrgDeletionFeatureFlag)
```
instead of getting a feature flag for a specific organization, I think it would be better here if we have a global feature flag. We either want hard deletes on the whole instance or we don't.

miveronese replied:
It makes sense

miveronese replied:
I am assuming that, in this case, the FF value should be fetched from the context?  Like this:
```
	flags := featureflag.FromContext(ctx)
	value := flags.GetBoolOr("org-deletion", false)
```

kopancek commented on client/web/src/org/backend.ts:
```diff
@@ -119,6 +124,33 @@ export function updateOrganization(id: Scalars['ID'], displayName: string): Prom
+/**
+ * WIP....
+ */
+export function hardDeleteOrganization(organization: Scalars['ID']): Promise<void> {
+    return requestGraphQL<HardDeleteOrganizationResult, HardDeleteOrganizationVariables>(
+        gql`
+           mutation HardDeleteOrganization($organization: ID!) {
+                hardDeleteOrganization(organization: $organization) {
+                    alwaysNil
+                }
+            }
+        `,
+        {
+            organization,
+        }
+    )
+    .pipe(
+        map(dataOrThrowErrors),
+        map(data => {
+            if (!data.deleteOrganization) { // WIP: find out what this does.
+                throw createInvalidGraphQLMutationResponseError('HardDeleteOrganization')
+            }
+        })
+    )
+    .toPromise()
+}
```
We are trying to migrate away from the `requestGraphql` to use React hooks `useQuery` or `useMutation` instead. See the latest changes from Pietro for an example: https://github.com/sourcegraph/sourcegraph/blob/main/client/web/src/user/settings/privacy/UserSettingsPrivacyPage.tsx#L32

miveronese commented on cmd/frontend/graphqlbackend/org_test.go:
```diff
@@ -431,3 +431,72 @@ func TestUnmarshalOrgID(t *testing.T) {
+	})
+
+	t.Run("Org is successfully deleted when feature flag is enabled", func(t *testing.T) {
+		featureFlags.GetOrgFeatureFlagFunc.SetDefaultReturn(true, nil)
```
@sourcegraph/cloud-saas ,wondering why line 479 (`featureFlags.GetOrgFeatureFlagFunc.SetDefaultReturn(true, nil)`)  is not working as in other tests that use the same approach. I mean: even when the bool passed is true, the  test fails because the flag is not enabled.... I checked all tests that use this same approach and all work fine except another one in this same file. On line 244, when changing the bool passed to `featureFlags.GetOrgFeatureFlagFunc.SetDefaultReturn( )` to false,  I would expect the test to fail, but it doesn't... 🤔

unknwon replied:
I did not dig deeper, but it looks to me the method you really want to mock is not the `GetOrgFeatureFlagFunc`.

```go
	flags := featureflag.FromContext(ctx)
	orgHardDeletionEnabled := flags.GetBoolOr("org-hard-deletion", false)
```

unknwon commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +284,29 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
+		return nil, errors.New("hard deleting organization is not supported")
+	}
+
+	if err := database.Orgs(r.db).HardDelete(ctx, orgID); err != nil {
```
You likely want this otherwise you'll hit a panic once the feature flags mocking is fixed.

```suggestion
	if err := r.db.Orgs().HardDelete(ctx, orgID); err != nil {
```

miveronese replied:
thanks, @unknwon . That fixed the issue and saved from a rabbit hole. 🎉

unknwon commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +283,31 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
 	return &OrgResolver{db: r.db, org: updatedOrg}, nil
 }
 
+// RemoveOrganization removes an org and all resources associated to it. Depending on the flag value, it will hard delete or soft delete the org.
```
>Depending on the flag value, it will hard delete or soft delete the org.

I do not see anywhere we are doing soft-delete? From what I understood, this endpoint is solely about hard deleting the organization, is that correct?

unknwon commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +283,31 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
+	}
+
+	orgDeletionFlag, err := r.db.FeatureFlags().GetFeatureFlag(ctx, "org-deletion")
+	if err != nil || orgDeletionFlag == nil || !orgDeletionFlag.Bool.Value {
+		return nil, errors.New("hard deleting organization is not supported")
+	}
```
We are currently swallowing the underlying error:

```suggestion
	if err != nil {
		return nil, err
	} else if orgDeletionFlag == nil || !orgDeletionFlag.Bool.Value {
		return nil, errors.New("hard deleting organization is not supported")
	}
```

unknwon commented on cmd/frontend/graphqlbackend/org_test.go:
```diff
@@ -431,3 +433,110 @@ func TestUnmarshalOrgID(t *testing.T) {
+	mockedOrg := types.Org{ID: 1, Name: "acme"}
+	orgIDString := string(MarshalOrgID(mockedOrg.ID))
+
+	mockedFeatureFlag := featureflag.FeatureFlag{Name: "org-deletion", Bool: &featureflag.FeatureFlagBool{Value: false}, Rollout: nil, CreatedAt: time.Now(), UpdatedAt: time.Now(), DeletedAt: nil}
```
nit: could you break this struct initialization into multiple lines? (i.e. each field on its own line). This long is getting a bit long.

unknwon commented on cmd/frontend/graphqlbackend/org_test.go:
```diff
@@ -431,3 +433,110 @@ func TestUnmarshalOrgID(t *testing.T) {
+
+	mockedFeatureFlag := featureflag.FeatureFlag{Name: "org-deletion", Bool: &featureflag.FeatureFlagBool{Value: false}, Rollout: nil, CreatedAt: time.Now(), UpdatedAt: time.Now(), DeletedAt: nil}
+	featureFlags := database.NewMockFeatureFlagStore()
+	featureFlags.GetFeatureFlagFunc.SetDefaultHook(func(ctx context.Context, flagName string) (*featureflag.FeatureFlag, error) {
+		return &mockedFeatureFlag, nil
+	})
```
Simpler way:

```suggestion
	featureFlags.GetFeatureFlagFunc.SetDefaultReturn(&mockedFeatureFlag, nil)
```

unknwon commented on cmd/frontend/graphqlbackend/org_test.go:
```diff
@@ -431,3 +433,110 @@ func TestUnmarshalOrgID(t *testing.T) {
+		})
+	})
+
+	mockedFeatureFlag2 := featureflag.FeatureFlag{Name: "org-deletion", Bool: &featureflag.FeatureFlagBool{Value: true}, Rollout: nil, CreatedAt: time.Now(), UpdatedAt: time.Now(), DeletedAt: nil}
+	featureFlags.GetFeatureFlagFunc.SetDefaultHook(func(ctx context.Context, flagName string) (*featureflag.FeatureFlag, error) {
+		return &mockedFeatureFlag2, nil
+	})
```
ditto

unknwon commented on internal/database/orgs.go:
```diff
@@ -308,3 +310,47 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+		err = tx.Done(err)
+	}()
+
+	if _, err := tx.Handle().DB().ExecContext(ctx, "DELETE FROM names WHERE org_id=$1", id); err != nil {
+		return err
+	}
+
+	if _, err := tx.Handle().DB().ExecContext(ctx, "DELETE FROM org_members WHERE org_id=$1", id); err != nil {
+		return err
+	}
+
+	if _, err := tx.Handle().DB().ExecContext(ctx, "DELETE FROM notebooks WHERE namespace_org_id=$1", id); err != nil {
+		return err
+	}
+
+	// For On-Prem, notebooks are enabled through settings, so this entry needs to be removed before an org is deleted.
+	if !envvar.SourcegraphDotComMode() {
+		if _, err := tx.Handle().DB().ExecContext(ctx, "DELETE FROM settings WHERE org_id=$1", id); err != nil {
+			return err
+		}
+	}
```
I haven't checked each, but if we have defined `ON DELETE CASCADE` on the [database schema](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/schema.md), technically we do not need to do manual deletions. I only checked `names` table which does have this property.

miveronese replied:
Yes,  you are right about the `names` table, @unknwon . It has a ON DELETE CASCADE (my bad :)) , but org_members doesn't, so I will keep the ones that don't have this property.

unknwon commented on internal/database/orgs.go:
```diff
@@ -308,3 +310,47 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+		return err
+	}
+
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows == 0 {
+		return &OrgNotFoundError{fmt.Sprintf("id %d", id)}
+	}
```
Practically, what is the rationale of returning an error if I want to delete an organization that simply does not exist?

Also, is this checkpoint too late here? Because if this organization with the ID does not exist, how could other resources binding to the organization exist?

unknwon commented on internal/database/orgs_test.go:
```diff
@@ -99,6 +99,59 @@ func TestOrgs_Delete(t *testing.T) {
+
+	displayName := "org1"
+	org, err := Orgs(db).Create(ctx, "org1", &displayName)
+	if err != nil {
+		t.Fatal(err)
+	}
```
nit: We have this test helper
```suggestion
	require.NoError(t, err)
```

unknwon reviewed (changes requested):
Overall LGTM 👍 

Left some comments, request changes because I want to take a last look before merging.

unknwon commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +283,33 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
+	Organization graphql.ID
+}) (*EmptyResponse, error) {
+	if !envvar.SourcegraphDotComMode() {
+		return nil, errors.New("hard deleting organization is not supported")
```
nit: this would the error message a bit more clear, since currently it is identical to the error message of the feature flag not enabled.
```suggestion
		return nil, errors.New("hard deleting organization is only supported on Sourcegraph.com")
```

unknwon commented on internal/database/orgs.go:
```diff
@@ -308,3 +309,54 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+
+func (o *orgStore) HardDelete(ctx context.Context, id int32) (err error) {
+	// Check if the org exists even if it has been previously soft deleted
+	orgs, err := o.getBySQL(ctx, "WHERE id=$1 LIMIT 1", id)
+	if err != nil {
+		return err
+	}
```
99% sure that if there is no record exists, `getBySQL` returns error `sql.ErrNoRows`, so what you actually want would be:
```suggestion
	orgs, err := o.getBySQL(ctx, "WHERE id=$1 LIMIT 1", id)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil
		}
		return err
	}
```

Ignore this if not the case.

miveronese replied:
`o. getBySQL`  returns an empty array when no record is found, so I will skip this suggestion.

unknwon reviewed (approved):
Nothing blocking, LGTM 👍

unknwon commented:
Never tried, but looks like we may be able to dismiss reviews. If you feel your other work are being blocked by merge of this PR and you're confident, let's dismiss and rely on post-merge feedback. Otherwise, let's wait for Milan.

<img width="939" alt="CleanShot 2022-03-01 at 11 46 11@2x" src="https://user-images.githubusercontent.com/2946214/156101006-16f1aa8f-865f-4f8b-8c66-68ec8d87acdd.png">

unknwon commented:
BTW CI is not passing.

rafax commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -283,6 +283,33 @@ func (r *schemaResolver) UpdateOrganization(ctx context.Context, args *struct {
+	orgDeletionFlag, err := r.db.FeatureFlags().GetFeatureFlag(ctx, "org-deletion")
+	if err != nil {
+		return nil, err
+	} else if orgDeletionFlag == nil || !orgDeletionFlag.Bool.Value {
```
nit: do you really need the else here? if the previous condition is true, we will exit the function, so I think we can remove the else to follow the Go style https://go.dev/doc/effective_go#if

miveronese replied:
Good point, @rafax ! I will update this part.

rafax commented on internal/database/orgs.go:
```diff
@@ -308,3 +309,54 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+		err = tx.Done(err)
+	}()
+
+	if _, err := tx.Handle().DB().ExecContext(ctx, "DELETE FROM org_members WHERE org_id=$1", id); err != nil {
```
nit: would it make sense to add the delete statements to an array, and iterate over it? with the amount of repetition right now, it would be hard to spot a missing ! or any other small mistakes

rafax commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -99,6 +99,10 @@ type Mutation {
     """
     deleteOrganization(organization: ID!): EmptyResponse
     """
+    Removes an organization.
```
Can you provide more context here for users of this API- how is "removeOrganization" different from "deleteOrganization" above? Should one be called "hardDelete"? Or, should we mention one of those marks the org a as deleted, and the other cleans up dependencies and deletes?

miveronese replied:
I update the description of both methods (removeOrganization and deleteOrganization) @rafax , but haven't changed the name of the first to "hardDelete" -- initially, it was named this way, but @kopancek  noted that a better name would be "removeOrganization" as this is the convention we use on other graphql resolvers.

unknwon replied:
Forgot the fact we already have `deleteOrganization`, then I strongly think we should just add `hardDelete` option instead of a new method, like for deleting users: 

https://github.com/sourcegraph/sourcegraph/blob/ff97e1deece731ead34679cc490b4adc2aa97f11/cmd/frontend/graphqlbackend/schema.graphql#L210

miveronese replied:
@unknwon , I've following your suggestion to use `deleteOrganization` instead of `removeOrganization` and am wondering if it is ok to leave that first function (`deleteOrganization`) where it is now [(in graphqlbackend/site_admin.go](https://github.com/sourcegraph/sourcegraph/blob/ff97e1deece731ead34679cc490b4adc2aa97f11/cmd/frontend/graphqlbackend/site_admin.go#L102))  or if it would make more sense to move it to [/graphqlbackend/org.go](https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/graphqlbackend/org.go)

unknwon reviewed (changes requested):
Request changes because I think we need [authz check](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@19c853f071131e0598ba16a931570484921ce34c/-/blob/cmd/frontend/graphqlbackend/org.go?L143), otherwise anyone is able to delete any org on cloud.

rafax commented on internal/database/orgs.go:
```diff
@@ -308,3 +309,47 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+	// The order of deletion matters, that is why we iterate over an array instead of a map, which is unordered.
+	tables := []string{"org_members", "org_invitations", "registry_extensions", "saved_searches", "notebooks", "settings", "orgs"}
+	for _, t := range tables {
+		query := fmt.Sprintf("DELETE FROM %s WHERE %s=%v", t, tablesAndKeys[t], id)
```
nit: inline query? but up-to-you

unknwon replied:
FYI, `fmt.Sprintf` is vulnerable to SQL injection, though I don't think it matters here as values are essentially hard-coded in code. But could be better to put a [security comment](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+SECURITY%3A&patternType=literal) saying be cautious about changing lines here.

The "correct" way is:

https://github.com/sourcegraph/sourcegraph/blob/9b506745ca5548a85eb76ec688ead71146e78bb1/internal/database/repos.go#L1598

rafax reviewed (changes requested):
I think we need a bit more authz here

rafax commented on cmd/frontend/graphqlbackend/org_test.go:
```diff
@@ -431,3 +433,124 @@ func TestUnmarshalOrgID(t *testing.T) {
+	ctx := actor.WithActor(context.Background(), &actor.Actor{UID: 1})
+	orig := envvar.SourcegraphDotComMode()
+
+	t.Run("Returns an error when in DotComMode but feature flag is not enabled", func(t *testing.T) {
```
Add a case for user trying to delete an org they are not a member of?

miveronese commented:
@unknwon and @rafax , the PR has been updated based on your last suggestions. 
- Deleted the `removeOrganization` function and moved its logic to  the `deleteOrganization`, so now it can handle both soft and hard deletions.
- Added a check to allow only org members to hard deleting their org
- Updated the tests accordingly
(Fingers crossed :D )

unknwon reviewed (approved):
Nothing blocking from me, thanks for the followup!

miveronese commented:
> Nothing blocking from me, thanks for the followup!

Yes!

miveronese commented:
Trying to figure out what the error showed by buildkite/sourcegraph ("Misc linters") mean... I run the linters several times locally and didn't see any errors... 🤔  Any suggestions?

unknwon commented:
> Trying to figure out what the error showed by buildkite/sourcegraph ("Misc linters") mean... I run the linters several times locally and didn't see any errors... 🤔 Any suggestions?

Looks like some copy paste error?  https://buildkite.com/sourcegraph/sourcegraph/builds/135078#ba83ca0f-6dea-45fc-8106-9869c7138d1b/769-770

```diff
-// SetDefaultReturn calls SetDefaultDefaultHook with a function that returns
-// the given values.
+// SetDefaultReturn calls SetDefaultHook with a function that returns the
+// given values.
```

Maybe try regenerate the mock?

unknwon commented:
Your build is now green! Let's ask @kopancek and @rafax to give a final look.

kopancek reviewed (approved):
LGTM!

rafax reviewed (approved):
A few non-blocking comments, LGTM otherwise.

miveronese commented on cmd/frontend/graphqlbackend/site_admin.go:
```diff
@@ -101,20 +102,48 @@ func (r *schemaResolver) DeleteUser(ctx context.Context, args *struct {
+			return nil, err
+		}
+	} else {
+		// 🚨 SECURITY: For On-premise, only site admins can soft delete orgs.
```
Your observation makes sense, @rafax . By the way, do we want to allow soft delete for Cloud too?

miveronese replied:
Here's the commit with the helper and a check for soft delete + updated tests: https://github.com/sourcegraph/sourcegraph/pull/31528/commits/c4dc1a41e5993580a653120a5025e065db1c91c0

unknwon commented on internal/database/orgs.go:
```diff
@@ -308,3 +309,47 @@ func (o *orgStore) Delete(ctx context.Context, id int32) (err error) {
+
+	// Some tables that reference the "orgs" table do not have ON DELETE CASCADE set, so we need to manually delete their entries before
+	// hard deleting an org.
+	tablesAndKeys := map[string]string{
```
Makes total sense to me!
